### DISCUSSION TITLE
QRange element

### DISCUSSION THREAD

**Discussion by frankvp11** [zauberzeug/nicegui#1191] [Ideas / Feature Requests]
Conveniently, my project needs exactly this: https://quasar.dev/vue-components/range/
I've searched through the PRs, issues and previous discussions and found nothing related to this. 
It would be cool if this could be added into nicegui - perhaps as either it's own element or as an add-on to the slider element.

**Comment by rodja** [comment]
Thanks for the feature request @frankvp11. Normally you should be able to already use it in some kind of way by writing

```py
ui.element('q-range').props('min=0 max=100 label drag-range')
```
But the UI seems to be a bit broken. Who wants to investigate this and maybe even create the real `ui.range` element?

**Comment by falkoschindler** [comment]
There has been discussion about QRange over at Discord: https://discord.com/channels/1089836369431498784/1093461421448237067

The problem with @rodja's solution (https://github.com/zauberzeug/nicegui/discussions/1191#discussioncomment-6484315) is that we need to initialize the `model-value` with two values `min` and `max`, which doesn't seem to work with the `.props` method. But we can define the `_props` dictionary directly:

```py
qrange = ui.element('q-range').classes('m-4')
qrange._props = {
    'min': 0,
    'max': 12,
    'step': 1,
    'model-value': {'min': 3, 'max': 9},
    'snap': True,
    'label-always': True,
}
qrange.on('update:model-value', lambda msg: print(msg))
```

In analogy with QRange it might be reasonable to introduce a separate UI element `ui.range` instead of extending `ui.slider`.

**Reply by frankvp11** [reply]
This solved my problem. I also think it might be a good idea to make a new ui element for this. Nonetheless, thank you very much for the help

**Reply by DarkCTO** [reply]
This is great! I will be doing the same with QRating

**Comment by frankvp11** [comment]
One thing that I noticed using @falkoschindler's fix, is that when I initialize it to be the max (that I want), then change it to a lower value (any value works) then back up to the max, the label-always part says that max (12), but in the events it doesn't update to the max again - something smaller always. 
The code to test this is the same as falkoschindler's, just with model-value max equal to 12. Then do as I instructed

**Reply by falkoschindler** [reply]
Wow, thanks for spotting this, @frankvp11!

In fact, whatever the initial value is set to, the event is never fired for this initial value. This is because I forgot to send the new value back to the client:

```py
qrange = ui.element('q-range').classes('m-4 w-64')
qrange._props = {
    'min': 0,
    'max': 12,
    'step': 1,
    'model-value': {'min': 3, 'max': 9},
    'snap': True,
    'label-always': True,
}

def handle_change(e: events.GenericEventArguments):
    print(e.args)
    qrange._props['model-value'] = e.args
    qrange.update()

qrange.on('update:model-value', handle_change)
```

The reason why this update is required, is a bit hard to explain. But basically the client only signals a user interaction and the server needs to send the new state back to the client.

**Reply by frankvp11** [reply]
I tested it locally and it worked - thank you.

**Comment by falkoschindler** [comment]
Let's keep this discussion open. A native `ui.range` element would still be a nice feature. 🙂

**Comment by mohsenminaei** [comment]
Add me to the list of people that needs a "range" component. Also thanks for providing a workaround.

**Reply by stdusr** [reply]
#MeToo

**Reply by falkoschindler** [reply]
We hear you. 🙂

By the way: Even though it's not a strict criterion, we occasionally review the list of feature requests sorted by upvotes. So, feel free to vote for the features you'd like to see in upcoming releases.

**Comment by sonnygeorge** [comment]
+1 to this feature request :)

**Comment by falkoschindler** [comment]
PR #2762 will be released in version 1.4.20.

**Comment by mohsenminaei** [comment]
Thank you!
